### ISSUE TITLE
Mac OS as a host OS?  🙃

### ISSUE THREAD

**Issue by dcorbe** [89luca89/distrobox#36]
Is this an effort worth exploring?

**Comment by 89luca89** [comment]
Yea I would like in the future to support it, what blocks me for now is the host integration

On MacOS podman runs in a VM, and distrobox integrates with the container's host, which in this case will be the VM, not the MacOS host
So it becomes quite a limited experience, without exports of stuff or graphical apps integration 

I'm gonna keep this issue for the future and mark it ad blocked until we either find a solution or we create a piece of documentation that explains the situation and does not confuse the users

Ironically I think (not tried yet) that distrobox on WSL2 could retain some real host integration seeing how the linux VM is tightly integrated with the Windows host :shrug:

**Comment by DanielTolentino** [comment]
> Yea I would like in the future to support it, what blocks me for now is the host integration
> 
> On MacOS podman runs in a VM, and distrobox integrates with the container's host, which in this case will be the VM, not the MacOS host So it becomes quite a limited experience, without exports of stuff or graphical apps integration
> 
> I'm gonna keep this issue for the future and mark it ad blocked until we either find a solution or we create a piece of documentation that explains the situation and does not confuse the users
> 
> Ironically I think (not tried yet) that distrobox on WSL2 could retain some real host integration seeing how the linux VM is tightly integrated with the Windows host 🤷

Tried with an Ubuntu WSL install, got an error, I'll try other distros later.
![image](https://user-images.githubusercontent.com/5665358/146834423-108a3004-08ba-4b3b-a1a2-375646311bc0.png)

**Comment by 89luca89** [comment]
Yea as said I didn't try 😂
Anyway that seems an incompatibility between wsl2 and podman probably? Maybe there are some guides for it

Anyway I'd stick this conversation to macos and in case there is interest open a new one for WSL2
(I doubt WSL1 will work)

**Comment by goodevilgenius** [comment]
I just discovered distrobox, and this was the first thought I had. I have to use a Mac for work, but would love to be able to use this for a bunch of stuff.

Although podman might be difficult, maybe docker would work better, due to it's somewhat better integration into the host with HyperKit (hypervisor).

I'm not really sure of the specifics of how that works, but it might be a consideration.

**Comment by dcorbe** [comment]
> On Dec 29, 2021, at 09:26, Dan Jones ***@***.***> wrote:
> 
> 
> Although podman might be difficult, maybe docker would work better, due to it's somewhat better integration into the host with HyperKit (hypervisor).

Docker also runs in a VM on Mac OS

If the goal is to containerize and not isolate, why not consider distributing a light-weight Kubernetes distribution with this project?  Something like k3s or Minikube would work perfectly for this and it has the added advantage of being able to run on Windows, Mac OS and Linux.

**Comment by 89luca89** [comment]
All of them share the VM problem anyway, podman docker k3s, minikube all work inside a linux VM
I think it would be doable as a way to have something to have pet containers for dev environments, but I see it difficult to have the level of integration that can be achieved on a Linux host

**Comment by dcorbe** [comment]
Do you think this might be closer to what we need?

https://github.com/gramineproject/gramine#what-is-gramine

**Comment by 89luca89** [comment]
Let's follow this as it's a blocker for now: https://github.com/containers/podman/issues/8016

**Comment by WoMspace** [comment]
My two cents are that I'd love to have distrobox on macOS regardless of the lack of integration, as an easy way to automatically spin up lightweight VMs for various uses.
I love WSL on windows and just installed distrobox on a linux machine and it's amazing there too, would love to have some analogue on macOS

**Comment by 89luca89** [comment]
Well the point of toolbox/distrobox type of software is the tight integration with the host
If I cannot provide this integration it's just like using podman/docker

On MacOS right now I'm not even sure graphical apps can work inside distrobox with podman-machine

**Comment by kevindurb** [comment]
While i totally agree that tight integration with the host is what makes distrobox amazing i think there is some amount of usage for macos... I wanted to work on my personal distrobox container today and didnt have my linux machine near by and ended up doing `podman build...` and `podman run...` to be able to mess with it a little... but the problem is i still didnt have my user setup correctly as distrobox automagically does for me... but i also would rather not add a massive amount of bloat to this project just to support a weird usecase... this project is so fantastic already id hate to taint it with an extra chunk of bloat to have to maintain...

**Comment by JakeHL** [comment]
It might be worth investigation [colima](https://github.com/abiosoft/colima) and/or [lima](https://github.com/lima-vm/lima) for macos support rather than podman as it offers tighter integration with the OS out of the box. (Still a vm however)

**Comment by 89luca89** [comment]
@JakeHL will investigate, I don't have a Mac so it might take some time :smile:

**Comment by JakeHL** [comment]
Let me know if I can be of help. I'll take a look in the meantime. Would love to see this up and running

**Comment by jacobnollette** [comment]
This would be increadible - mac os is the only option at work

**Comment by stevo-knievo** [comment]
1+

**Comment by aral** [comment]
> Let's follow this as it's a blocker for now: [containers/podman#8016](https://github.com/containers/podman/issues/8016)

Heads up, looks like that blocker is now closed on the pod man side of things: https://github.com/containers/podman/issues/8016 :)

**Comment by jakehl** [comment]
I think podman honestly just over complicates things. Lima should be perfect for something like this - it allows for invoking commands inside a VM, sharing state between host and VM etc. Honestly its not far from doing what distrobox already does. Ive started to play with it to create a distrobox like experience and its been pretty great. Ive not tried integrating actual distrobox yet though.

**Comment by aral** [comment]
@jakehl Hey Jake, is your work available somewhere to play with? Sounds very useful :)

**Comment by jakehl** [comment]
Its not, its just crappy shell aliases and functions to try and recreate the behavior. If I can get some time over the holiday break I'll tidy it up and push it.

**Comment by kevindurb** [comment]
ok so i started messing with getting distrobox installed on macos, the install script uses `install` which is a different version from what is on macos so i did `brew install coreutils` and then replaced `install` with `ginstall` to get all the distrobox scripts put into place, that worked... 

Next problem is when i go to run `distrobox create` I get this:
```
Error: failed to connect: dial tcp [::1]:51787: connect: connection refused

An error occurred

An error occurred
```

**Comment by kevindurb** [comment]
lol that error was because I didnt have my podman machine running 🤦🏻‍♂️...  more errors on create using podman `Error: statfs /Users/kevindurbin/.local/bin/distrobox-export: no such file or directory`

**Comment by kevindurb** [comment]
container stays stuck in `CREATED`... I think its due to performance of mounting all of `~` with macos docker...

**Comment by 89luca89** [comment]
> lol that error was because I didnt have my podman machine running 🤦🏻‍♂️... more errors on create using podman `Error: statfs /Users/kevindurbin/.local/bin/distrobox-export: no such file or directory`... testing docker...

Yes, this is because some parts of distrobox (init, host-exec, export) are mounted inside the container itself
In case of lima and other VM based solutions (which includes docker), the path of distrobox is not present, nor the same as inside the VM
So this will create errors

**Comment by kevindurb** [comment]
Podman on macos is also a VM based solution. I dont know of any container system that doesnt run on a vm on macos.

**Comment by jakehl** [comment]
> Yes, this is because some parts of distrobox (init, host-exec, export) are mounted inside the container itself...

What is the path it should be mounted at? Lima has great mounting options to share filesystems between host/guest. I'll see if I can get it working.

**Comment by kevindurb** [comment]
looks like theres an extensive list of volumes that get added, if you look at distrobox-create generate_command() function you can see it add all the volumes in. heres the initial list, i assume that `/tmp` is the one associated with the error, gonna attempt to just not mount this... but i'm also wondering if theres something broken with macos and `rslave` mounts...?
```
		--volume /:/run/host:rslave
		--volume /dev:/dev:rslave
		--volume /sys:/sys:rslave
		--volume /tmp:/tmp:rslave
		--volume \"${distrobox_entrypoint_path}\":/usr/bin/entrypoint:ro
		--volume \"${distrobox_export_path}\":/usr/bin/distrobox-export:ro
		--volume \"${distrobox_hostexec_path}\":/usr/bin/distrobox-host-exec:ro
		--volume \"${container_user_home}\":\"${container_user_home}\":rslave"
```

**Comment by kevindurb** [comment]
yea it seems like theres something macos doesnt like about rslave mounts... i have no clue what an rslave mount is so gonna do some research to see what this does and maybe theres an alternative for macos

**Comment by kevindurb** [comment]
https://docs.docker.com/storage/bind-mounts/#configure-bind-propagation
> Mount propagation doesn’t work with Docker Desktop.
Well theres that answer lol

**Comment by kevindurb** [comment]
overall that means that if you remove some mounts and change the home mount to be a "regular" (`shared`?) mount and change the installer to use ginstall then you can have a limited functionality working distrobox on macos 🎉

**Comment by kevindurb** [comment]
PR to get the installer working for macos https://github.com/89luca89/distrobox/pull/615

**Comment by kevindurb** [comment]
Updated my pr to also add the changes to get `create` working. With those changes distrobox now works on macos in a more limited capacity (no /dev for device support, etc)

**Comment by 89luca89** [comment]
I'm glad there is progress in this issue, thanks @kevindurb :smile: 
Hopefully I'll find a way to run macos in vm (laptop/cloud) and do some help :+1:

**Comment by pythoninthegrass** [comment]
Lost my original draft, but I might be able to get you a 2017 x86 MacBook Air from work @89luca89.

Or I can test whatever you want on either a [runner](https://github.com/actions/runner-images/blob/main/images/macos/macos-12-Readme.md) or locally — have access to a number of x86 and arm64 Macs.

**Comment by FedX-sudo** [comment]
Potentially stupid question: How feasible is it to add MacOS support using something like [Docker-OSX](https://github.com/sickcodes/Docker-OSX)?

**Comment by Soromeister** [comment]
Any updates on this?

**Comment by osalbahr** [comment]
> I daily drive an m1 macbook, Let me know if I can be of help. I'll take a look in the meantime. Would love to see this up and running

Same here, except I daily drive an Intel-based MacBook. Would love to help port distrobox to macOS, let me know if there is anything I should try. I’d like to have a distrobox with development tools in the same seamless way. I don’t know if being able to export is a main feature.

**Comment by osalbahr** [comment]
I created Homebrew/discussions#4603 and hopefully getting a fully functional `brew install distrobox` on Linux would make porting to macOS easier

**Comment by redthing1** [comment]
I'm very interested in this too

**Comment by tl87** [comment]
I would also really like for this to be a reality 👍🏻

**Comment by kevindurb** [comment]
If anything it might be cool to add a `lima.yaml` to this repo that does the mount and installs distrobox automatically

**Comment by Sahil-Gulihar** [comment]
Has it been resolved?

**Comment by osalbahr** [comment]
> Has it been resolved?

As of now, [Homebrew lists distrobox as Linux-only](https://formulae.brew.sh/formula/distrobox), so I guess not.

This was also confirmed in March, 2024, in a Homebrew discussion: https://github.com/orgs/Homebrew/discussions/5190 (at least on M1, no official support for distrobox on macOS). To be clear, this is not Homebrew's issue to fix, it's just an easy status check.

@Sahil-Gulihar This issue remains open if you would like to work on it. You'll probably have to do some form of a limited-feature, container-in-container solution, which I don't know the use case for that compared to directly using containers (what I personally ended up doing, it's a good-enough workaround for my use).

- Relevant discussion: https://github.com/89luca89/distrobox/issues/1021

(I'm no longer interested in actively working on it, but would be happy to try to help if you run into an issue)

**Comment by redthing1** [comment]
To anyone who's interested in this: I wanted something like distrobox for a long time, after switching to a mac.
So I wrote my own minimalist implementation of something like distrobox. You define your linux box environment using a dockerfile (samples provided), then you can seamlessly mount your home/user directories in, so you can work on them directly but from a linux terminal.

Specifically designed and tested on mac but also will work anywhere supporting docker/podman. If you want to use docker instead of podman just Ctrl+F replace, it works with both.

I hope this is useful to others who switched to mac and miss distrobox!

https://github.com/redthing1/mim

**Comment by Sahil-Gulihar** [comment]
> Specifically designed and tested on mac

Does it solve the issue of accessing GUI apps from host system to the docker/podman?

**Comment by ericcurtin** [comment]
FWIW I did manually get this working using

podman machine

and

podman run -e DISPLAY=some-ip-address:0 container-image

and

xquartz

using this guide:

https://gist.github.com/sorny/969fe55d85c9b0035b0109a31cbcb088

so it's possible at least.

**Comment by corgijan** [comment]
Im also very interested, would make work such a breeze

**Comment by ericcurtin** [comment]
One thing I will say is it work be nice if "/User" was added as a shared volume by default, on macOS/podman-machine it's needed to share $HOME :

distrobox create --image fedora:41 --volume /Users:/Users

but I would recommend not checking "if on macOS" because when running distrobox in podman-machine, the platform is technically Linux.

**Comment by kevindurb** [comment]
The distro box scripts are still running on the macOS host so using "if macOS" is perfectly fine. But the other side of that problem is that by default not all of /User is shared into podman machine by default, only the current user home

**Comment by kevindurb** [comment]
FYI macOS on podman is working with just one flag passed on create. I haven't tested docker or Lima recently but podman with podman machine is working. https://github.com/89luca89/distrobox/pull/615#issuecomment-2752080322

**Comment by ericcurtin** [comment]
I execute distrobox inside podman-machine I have reasons for that, so both are valid use cases. I think sharing all of "/Users" if the directory exists is not a bad idea, permissions will still protect against directories that are not supposed to be rwx. Usernames may not always match up between macOS inside podman-machine and inside the podman container.

**Comment by ericcurtin** [comment]
Yup podman-machine is a Linux VM which automatically shares `/Users` for convenience, I can't see the negative of sharing `/Users` if it exists

**Comment by svenjacobs** [comment]
Apple just announced [container](https://github.com/apple/container). Could this be a possible backend for Distrobox on macOS?

**Comment by 89luca89** [comment]
> Apple just announced [container](https://github.com/apple/container). Could this be a possible backend for Distrobox on macOS?

it's interesting but it still poses the same limitations

would audio/video/gpu work? will I be able to pass devices to the container?
if not, then 50+% of what distrobox is useful for, won't work, and I have no intention to release something that works 50% and then deal with issues related to that 🤷

**Comment by ericcurtin** [comment]
> > Apple just announced [container](https://github.com/apple/container). Could this be a possible backend for Distrobox on macOS?
> 
> it's interesting but it still poses the same limitations
> 
> would audio/video/gpu work? will I be able to pass devices to the container? if not, then 50+% of what distrobox is useful for, won't work, and I have no intention to release something that works 50% and then deal with issues related to that 🤷

krunkit is the only VMM that allows GPU passthrough AFAIK today. And it's pretty neat.

**Comment by 89luca89** [comment]
> > > Apple just announced [container](https://github.com/apple/container). Could this be a possible backend for Distrobox on macOS?
> > 
> > 
> > it's interesting but it still poses the same limitations
> > would audio/video/gpu work? will I be able to pass devices to the container? if not, then 50+% of what distrobox is useful for, won't work, and I have no intention to release something that works 50% and then deal with issues related to that 🤷
> 
> krunkit is the only VMM that allows GPU passthrough AFAIK today. And it's pretty neat.

interesting indeed, but distrobox would first need some container manager to use it, and setup all the video/audio output devices too, in order to have something that would be feature-par with what linux can do

**Comment by osalbahr** [comment]
I just wiped my disk and re-installed macOS on my 2017 MacBook Pro. I still get the same errors. At the bottom are all the commands I ran:

```console
% distrobox create
Image registry.fedoraproject.org/fedora-toolbox:latest not found.
Do you want to pull the image now? [Y/n]: 
Trying to pull registry.fedoraproject.org/fedora-toolbox:latest...
Getting image source signatures
Copying blob sha256:d2e595eb6ec1ddc83ff04b8fbc3e9a616af7da8d3d2aebdf3ac09358e0153e8d
Copying config sha256:503805ef18c7b6e0c7119ea83d5a501fea3cd2ac960d5105b4a534bffce9fbbc
Writing manifest to image destination
503805ef18c7b6e0c7119ea83d5a501fea3cd2ac960d5105b4a534bffce9fbbc
Creating 'my-distrobox' using image registry.fedoraproject.org/fedora-toolbox:latest	 [ OK ]
Distrobox 'my-distrobox' successfully created.
To enter, run:

distrobox enter my-distrobox

Error: 2 errors occurred:
	* copying to host: copier: put: error resolving "/tmp": open /tmp: too many levels of symbolic links
	* copying from container: io: read/write on closed pipe
% history
    1  caffeinate -d
    2  /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
    3  eval "$(/usr/local/bin/brew shellenv zsh)"
    4  brew install podman
    5  podman machine init
    6  podman machine start
    7  git clone https://github.com/89luca89/distrobox.git
    8  cd distrobox
    9  ./install
   10  export PATH="$PATH:/Users/admin/.local/bin"
   11  distrobox create
% 
```

**Comment by rnk** [comment]
I wanted to report I'm using distrobox on Mac successfully today. I used this janky custom install script that simply fetches the bash scripts from main directly: https://gist.github.com/gianlucamazza/f9b57d6796a97981908f7c2bbda706fc

I had to exclude TMPDIR from the [environment passthrough](https://github.com/89luca89/distrobox/blob/main/distrobox-enter#L442), since TMPDIR is set to some unique per-user value, like mine: TMPDIR=/var/folders/4s/tl73q6493knb2n_ym5gr00b80000gn/T/. This volume appears to not be passed through, so it doesn't exist, so applications (fish) that use TMPDIR fail. I got this error:
```
  error: Runtime path not available.
  error: Try deleting the directory /var/folders/4s/tl73q6493knb2n_ym5gr00b80000gn/T//fish.<$USER> and restarting fish.
```

**Comment by flatsiedatsie** [comment]
> I used this janky custom install script

The script worked perfectly here.

Heck, I'd say that running this on MacOS is a killer feature for me.

**Comment by aral** [comment]
> I wanted to report I'm using distrobox on Mac successfully today. I used this janky custom install script that simply fetches the bash scripts from main directly: https://gist.github.com/gianlucamazza/f9b57d6796a97981908f7c2bbda706fc

This is very cool indeed (and worked perfectly for me also). I wonder how difficult it would be to create a brew formula for it…
